src/plugins/kibana_utils/common/registry/create_registry.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Observable, Subject } from 'rxjs';
+import { Registry, Predicate } from './types';
+
+export const createRegistry = <T>(): Registry<T> => {
```

Comment:
Overall, I primarily want to make sure we're not solving a more complicated problem than we need to. There are some choices that I'm a bit unclear about. What is the advantage of using generators instead of simply returning arrays? Do we benefit from providing an id when adding items? Why provide separate observables for set and clear?

Reply:
I guess we discussed registries at our meeting already. But for posterity:

> What is the advantage of using generators instead of simply returning arrays? 

This implementation handles all our registry use cases.

> Do we benefit from providing an id when adding items?

Because in all registries (I believe) records can be identified by ID, and in some `uiRegistries` and all `kbn-interpreter` registries records are also accessed by ID. So, registries are effectively key-value stores.

> Why provide separate observables for set and clear?

¯\_(ツ)_/¯ Just one way of doing it.